### DISCUSSION TITLE
Recommended way to type hint method decorator

### DISCUSSION THREAD

**Discussion by hmc-cs-mdrissi** [microsoft/pyright#3679] [Q&A] [answered]
As this question appears still open, https://github.com/python/typing/discussions/946 how would you recommend type hinting a decorator similar to cache? Is there any good way to type hint a decorator that fully preserves call signature of method and non-method and just adds a couple extra attributes. Toy example,

```python
class TrackedCallable(Generic[P, R]):
  count: int
  def __call__(...) -> R:
    ...

def track_calls(f: Callable[P, R]) -> TrackedCallable[P, R]:
  ...

class Foo:
  @track_calls
  def bar(self, x):
    ...

@track_calls
def bar2(x):
  ...
```

**Comment by erictraut** [comment, marked as answer]
I don't think this is a use case that the authors of PEP 612 contemplated or designed for.

You could create separate decorators, one that is designed for instance or class methods and the other that is designed for regular functions and static methods. I realize that's not a satisfying answer, but I don't think there is a good answer here given the way `ParamSpec` is designed currently.

You could propose and champion a change to the type system if you feel strongly about solving this problem.

**Reply by hmc-cs-mdrissi** [reply]
For decorators I write myself I do agree that's best approach given current type system. I ended up hitting this today with 3rd party library decorator and being unsure how to write a stub file for it. Maybe best workaround is to alias it into two decorators. Using cache as an example do something like,

```python
from functools import cache

def method_cache(func: Callable[Concatenate[SelfT, P], R]) -> _lru_cache_wrapper[P, R]:
  return cache(func) # type: ignore
```

If I hit this many more times I'll consider type system change. My guess is if I picked something to champion in typing I'd pick other things though (probably more stubs for ml libraries).
